lib/client.js

```diff
@@ -1288,57 +1314,87 @@ MqttClient.prototype._onConnect = function (packet) {
-  })
-  outStore.on('error', function (err) {
-    that.removeListener('close', remove)
-    that.emit('error', err)
-  })
+  function startStreamProcess () {
+    var outStore = that.outgoingStore.createStream()
 
-  function remove () {
-    outStore.destroy()
-    outStore = null
-  }
+    that.once('close', remove)
+    outStore.on('error', function (err) {
+      that.storeProcessing = false
+      that.packetIdsDuringStoreProcessing = {}
+      that.removeListener('close', remove)
+      that.emit('error', err)
+    })
 
-  function storeDeliver () {
-    // edge case, we wrapped this twice
-    if (!outStore) {
-      return
+    function remove () {
+      outStore.destroy()
+      outStore = null
+      that.storeProcessing = false
+      that.packetIdsDuringStoreProcessing = {}
     }
 
-    var packet = outStore.read(1)
-    var cb
+    function storeDeliver () {
+      // edge case, we wrapped this twice
+      if (!outStore) {
+        return
+      }
+      that.storeProcessing = true
 
-    if (!packet) {
-      // read when data is available in the future
-      outStore.once('readable', storeDeliver)
-      return
-    }
+      var packet = outStore.read(1)
 
-    // Avoid unnecessary stream read operations when disconnected
-    if (!that.disconnecting && !that.reconnectTimer) {
-      cb = that.outgoing[packet.messageId]
-      that.outgoing[packet.messageId] = function (err, status) {
-        // Ensure that the original callback passed in to publish gets invoked
-        if (cb) {
-          cb(err, status)
-        }
+      var cb
+
+      if (!packet) {
+        // read when data is available in the future
+        outStore.once('readable', storeDeliver)
+        return
+      }
 
+      // Skip already processed store packets
+      if (that.packetIdsDuringStoreProcessing[packet.messageId]) {
         storeDeliver()
+        return
+      }
+
+      // Avoid unnecessary stream read operations when disconnected
+      if (!that.disconnecting && !that.reconnectTimer) {
+        cb = that.outgoing[packet.messageId]
+        that.outgoing[packet.messageId] = function (err, status) {
+          // Ensure that the original callback passed in to publish gets invoked
+          if (cb) {
+            cb(err, status)
+          }
+
+          storeDeliver()
+        }
+        that.packetIdsDuringStoreProcessing[packet.messageId] = true
+        that._sendPacket(packet)
+      } else if (outStore.destroy) {
+        outStore.destroy()
       }
-      that._sendPacket(packet)
-    } else if (outStore.destroy) {
-      outStore.destroy()
     }
-  }
 
+    outStore.on('end', function () {
+      var allProcessed = true
+      for (var id in that.packetIdsDuringStoreProcessing) {
+        if (!that.packetIdsDuringStoreProcessing[id]) {
+          allProcessed = false
+          break
+        }
+      }
+      if (allProcessed) {
+        that.packetIdsDuringStoreProcessing = {}
+        that.storeProcessing = false
```

Comment:
I think these two properties are better prefixed with a `_`, those are internal.

Reply:
Ah, yes it should be. I updated the commit now.